### DISCUSSION TITLE
Having a hard time connecting to Docker when running Test in Xcode

### DISCUSSION THREAD

**Discussion by jaysonng** [netreconlab/parse-server-swift#30] [General]
So I have an existing iOS project which I'm trying to run Tests on the new parse-server-swift docker setup I have running. 

Everything seems to be ok with the Docker setup but when I run the Test on XCode, I get an error 

`error: ParseError code=141 error=connect ECONNREFUSED 127.0.0.1:8081`

My setup is 

/projects/MyProject
/projects/parse-server-swift

I've also set MyProject working directory to parse-server-swift trying to get it to connect.

![Screenshot 2023-05-24 at 6 00 38 PM](https://github.com/netreconlab/parse-server-swift/assets/440445/da88f6a5-73d5-4a72-a15c-8a5c536c3cf7)

When I access the url via localhost:8081/hello , the code works fine and I get a string back as expected.

What am I missing? Do I need to add a configuration somewhere that changes 127.0.0.1 to something else? 

Full error

```
Test Case '-[LokalityTests.LocationTests testHelloVapor]' started.
2023-05-24 19:25:21.880953+0800 Lokality[48461:457393] [connection] nw_socket_handle_socket_event [C1.1.1:2] Socket SO_ERROR [61: Connection refused]

2023-05-24 19:25:21.887477+0800 Lokality[48461:457393] [connection] nw_socket_handle_socket_event [C2.1.1:2] Socket SO_ERROR [61: Connection refused]
error: ParseError code=141 error=connect ECONNREFUSED 127.0.0.1:8081

/Users/jayson/Projects/Lokality/Shared/LokalityTests/LocationTests.swift:459: error: -[LokalityTests.LocationTests testHelloVapor] : XCTAssertNil failed: "testHelloVapor failed"

Test Case '-[LokalityTests.LocationTests testHelloVapor]' failed (0.387 seconds).
```

Tests

```
struct TestHelloVapor: ParseCloudable {

        //: Return type of your Cloud Function
        typealias ReturnType = String

        //: These are required by `ParseCloudable`
        var functionJobName: String = "hello"

        //: Parameters

    }
    
    func testHelloVapor() async {
        let testHelloVapor = await TestHelloVapor()

        do {
            let string = try await testHelloVapor.runFunction()
            print("Lokality: \(string)")
        } catch {
            print("error: \(error.localizedDescription)")
            XCTAssertNil("testHelloVapor failed")
        }

    }
```

Having a headache trying to figure it out on my own.

**Comment by cbaker6** [comment]
I’m not sure I understand what you are trying to do. ParseServerSwift connects directly to your node Parse Server, your apps should not attempt to connect to ParseServerSwift directly, instead, your app should connect to your node Parse Server as it normally would. The Node JS server will forward Requests to ParseServerSwift when needed and let it handle cloud code function and trigger requests. This functions exactly how JS Cloud Code functions.

**Comment by jaysonng** [comment]
Thats what i figured. But for some reason I can't get my app to connect to the parseserverswift cloud function or maybe im not understanding the error im getting. 

The econnrefused error I get means i have a configuration problem maybe?

error: ParseError code=141 error=connect ECONNREFUSED 127.0.0.1:8081

**Comment by cbaker6** [comment]
Your app should never connect to ParseServerSwift. ParseServerSwift should connect to your NodeJS ParseServer. Then your app should connect to your nodeJS Parse Server. Details about configuring are here: https://github.com/netreconlab/parse-server-swift#configure-parseserverswift-to-connect-to-your-parse-servers

**Reply by jaysonng** [reply]
yup I'm not connecting to ParseServerSwift. Im calling TestHelloVapor which is ParseCloudable when I get that error. 

```
 func testHelloVapor() async {
        let testHelloVapor = await TestHelloVapor()

        do {
            let string = try await testHelloVapor.runFunction()
            print("Lokality: \(string)")
        } catch {
            print("error: \(error.localizedDescription)")
            XCTAssertNil("testHelloVapor failed")
        }

    }
```
    
I'll keep looking into it and try to figure out why I'm getting that ECONNREFUSED error.

thanks

**Reply by cbaker6** [reply]
It sounds like you have a configuration issue. Do you see your webhooks in the Dashboard like the figure? https://github.com/netreconlab/parse-server-swift#in-docker What happens when you run `docker-compose up` when cloning the repo and running the Cloud Code directly from the playgrounds? https://github.com/netreconlab/Parse-Swift/blob/57ccb4be70feedbc289db65d4ff5732c3b9ef5ec/ParseSwift.playground/Pages/10%20-%20Cloud%20Code.xcplaygroundpage/Contents.swift#L59-L75

**Reply by jaysonng** [reply]
I did see the Webhooks on the Dashboard.

So I tried a fresh install and and ran the Playground Hello code as you suggested. It worked perfectly.

The problem came back when I switched PARSE_SERVER_DATABASE_URI to my own URI pointed to mongodb

I think the connection problem comes up there somewhere as last night I saw a "MasterKey required" error while debugging.  

Am I wrong in just switching PARSE_SERVER_DATABASE_URI  to my existing DB?

**Reply by cbaker6** [reply]
You have to make sure your URI contains the right credentials to connect to your MongoDB. The `parse-hipaa` image itself works with mongo as long as you configure it correctly. Here's the image config I know works with mongo: https://github.com/netreconlab/parse-hipaa/blob/0f2021e2f188cdd60dc7059ba23dedec3cfa1f48/docker-compose.mongo.yml#L15-L47

**Reply by jaysonng** [reply]
I know the URI works as I see all my Classes and I can see it all in Parse Dashboard. but yes, Ive been checking the credentials of the current one I use.

**Reply by cbaker6** [reply]
When you say Dashboard, I’m assuming you are talking about your own dashboard which may have the right configuration, but doesn’t mean you have parse server swift or parse-hipaa configured correctly. If you can see the dashboard using https://github.com/netreconlab/parse-server-swift/blob/f61336598f12fcc48c6c79f02787922420bd5b17/docker-compose.yml#L41-L48 and by accessing it through https://github.com/netreconlab/parse-server-swift#in-docker, then you have your configuration correct.

**Reply by jaysonng** [reply]
No, I mean Parse Dashboard which I access through localhost:1337/dashboard once Docker is up.

Btw, my mongodb is hosted over the cloud and not locally if that gives you any idea? Its hosted on mongodb.com and I just use the URL given there with username and password and then the db link.

**Reply by cbaker6** [reply]
> No, I mean Parse Dashboard which I access through localhost:1337/dashboard once Docker is up.

This means you are using the Dashboard through parse-hipaa which is Dashboard I give you here: https://github.com/netreconlab/parse-server-swift/discussions/30#discussioncomment-6003839

> Btw, my mongodb is hosted over the cloud and not locally if that gives you any idea? Its hosted on mongodb.com and I just use the URL given there with username and password and then the db link.


A hosted DB should be fine as long as you have the URI correct and the DB is allowing connections. My guess is your URI is incorrect. If your hosted DB requires a tls connection then you need configure the URI correctly to allow it, you may also need to configure the parse server with a certificate depending on your setup. I haven’t used Mongo in years, but it should be easy to google. A Heroku hosted Postgres example for a secure URI connection is here: https://github.com/netreconlab/parse-hipaa/blob/0f2021e2f188cdd60dc7059ba23dedec3cfa1f48/parse/index.js#L229. I describe this in detail (https://github.com/parse-community/docs/pull/731) in the Parse Server docs for Postgres, you can look for the Mongo equivalent: https://docs.parseplatform.org/parse-server/guide/#database

You should also look at your server logs in docker for additional info

**Reply by jaysonng** [reply]
From what I've read aroun, this error might be because mongodb isn't running locally. 

You think that might be it since  postgres is the default db?

My uri is correct I'm pretty sure as i do get to access it via the dashboard. Doesn't that mean Parse-Server has the correct access to the db ? 

And parse-server-swift getting to add new hooks into webhooks means that it is also connected to the db. 

I did look into the tls suggestion you mentioned but that one flew over my head. Ill have to read up on that some more. 

Ill continue in the morning and will see if i can install mongodb locally and somehow just work with that. Move over the data. Used up so much time already trying to connect my current setup.

**Reply by cbaker6** [reply]
> From what I've read aroun, this error might be because mongodb isn't running locally.

This doesn’t make sense to me, it shouldn’t matter if it’s local or remote, both communicate over a network

> You think that might be it since postgres is the default db?

Parse server uses Mongo by default. I set my images to use Postgres, the image works with both and Parse is designed to get mongo working easier than Postgres.

> And parse-server-swift getting to add new hooks into webhooks means that it is also connected to the db.

If you see the hooks along with your other Classes in the Dashboard, then your server is connected to the DB correctly. If your parse server is remote (sitting on Heroku/AWS) and ParseServerSwift is running on your local machine, then ParseServerSwift can talk to your remote parse server, but your remote parse server cannot talk to `ParseServerSwift` unless you bind it to a publicly accessible IP and configure `PARSE_SERVER_SWIFT_HOST_NAME` and `PARSE_SERVER_SWIFT_PORT` appropriately. Configuring the `env` variables is mentioned in the README: https://github.com/netreconlab/parse-server-swift#environment-variables

> Ill continue in the morning and will see if i can install mongodb locally and somehow just work with that. Move over the data. Used up so much time already trying to connect my current setup.

A local Mongo setup is easy and already in the code I’ve linked previously (plus the majority of parse examples online use mongo), simply adding https://github.com/netreconlab/parse-hipaa/blob/dd2f571aebce2cf213f5238bd10c7ee6614bd005/docker-compose.mongo.yml#L42-L50 and using the URI https://github.com/netreconlab/parse-hipaa/blob/dd2f571aebce2cf213f5238bd10c7ee6614bd005/docker-compose.mongo.yml#L15. Local/remote should not matter, docker networks all of the containers so they operate like everything is remote. your issue most likely means you misconfigured something else when switching your setup, particularly if the default from my repo works, but things break after your modifications. Maybe the `publicServerURL` is incorrect? I’m not able to understand your setup and what you’ve changed from what you’ve posted so far

**Reply by jaysonng** [reply]
The only thing i changed from the default docker setup where everything works is the databaseUri then I get the error. 

I've so far been testing locally, both docker setup and my own. Both get the error when I run xcode test. (But localhost:8081/hello works when accessed directly via an app i use for GET/POST - I get "Hello World" back)

Ill do more testing today. 

Thanks for the help.

**Reply by cbaker6** [reply]
>  I've so far been testing locally, both docker setup and my own. Both get the error when I run xcode test. (But localhost:8081/hello works when accessed directly via an app i use for GET/POST - I get "Hello World" back)

Xctest is intended to test the ParseServerSwift route (these are [vapor tests](https://docs.vapor.codes/advanced/testing/)) and not intended to test when ParseServerSwift is connected to a node parse server. There’s no way to tell if your Xctest environment is setup correctly. You should be connecting through playgrounds or your actual code app via a node parse server if you really want to test. You can try using a REST client, but there’s still no way for me to tell if you are sending JSON to ParseServerSwift correctly.

You should also look at your logs in Docker Desktop (both ParseServerSwift) and (node parse server), particularly when you hit your error. The logs are typically very informative will probably lead you towards the root cause of the issue you are facing.

You should also verify that you have pulled in the fixes from #28 and that you are using 5.7 https://github.com/netreconlab/parse-server-swift/blob/9ee15afdad40fe84bfc1950ffdabff98db97a53d/Dockerfile#L4

**Reply by jaysonng** [reply]
> You should be connecting through playgrounds or your actual code app via a node parse server if you really want to test.

Sorry, that's what I meant. I use your cloud code hello.runfunction playground example which works when I have your default Docker setup. and not working once I change to my mongodb databaseUri

![Screenshot 2023-05-27 at 11 43 43 PM](https://github.com/netreconlab/parse-server-swift/assets/440445/781ec676-aa67-44ef-9876-248dba882243)


> You should also look at your logs in Docker Desktop (both ParseServerSwift) and (node parse server), particularly when you hit your error. The logs are typically very informative will probably lead you towards the root cause of the issue you are facing.

```
error: Failed running cloud function hello for user undefined with:
  Input: {}
  Error: {"message":"connect ECONNREFUSED 127.0.0.1:8081","code":141} {"error":{"code":141,"message":"connect ECONNREFUSED 127.0.0.1:8081"},"functionName":"hello","params":{}}
error: connect ECONNREFUSED 127.0.0.1:8081 {"code":141,"stack":"Error: connect ECONNREFUSED 127.0.0.1:8081\n    at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1187:16)"}
```

that's the full error. I've been looking into TCPConnectWrap.afterConnect

**Reply by cbaker6** [reply]
I'm confident you've modified more than just `PARSE_SERVER_DATABASE_URI` which is different from what you've stated in https://github.com/netreconlab/parse-server-swift/discussions/30#discussioncomment-6016904, as `ECONNREFUSED 127.0.0.1:8081` doesn't make any sense for the original setup I've provided. This is probably due to you attempting to use the web hooks from `XCTest` and never removing them after.

From a docker standpoint, **none of the code I provided would ever** try communicate from a node Parse Server to ParseServerSwift through `127.0.0.1:8081`. Even in the [screenshots](https://github.com/netreconlab/parse-server-swift#in-docker) I provide, the web hooks show that node Parse Server communicates through `http://cloud-code:8081` which will directly use the local network docker creates to reach ParseServerSwift. I also [never bind](https://github.com/netreconlab/parse-server-swift/blob/9ee15afdad40fe84bfc1950ffdabff98db97a53d/docker-compose.yml#L66-L67) to `127.0.0.1:8081` as this is 1) Unnecessary, 2) Unsafe due to ParseServerSwift containing the `primaryKey` of your server (you basically want minimize public access to anything that has the `primaryKey`, besides your node Parse Server). Your error states it's trying to communicate to a port that is never opened or binded to which is why your connection is refused (or, you are no longer your ParseServerSwift in Xcode).

You will need to provide the proper information in order for me to assist you in the future. You will also need to ensure you are using docker properly. If you make a change to ParseServerSwift, you need to:

1. Delete the current web hooks in your dashboard by clicking the trash can next to each hook. This is to ensure you always are using the latest hooks and not using the XCTest hooks you've been attempting to use in the past
2. In your terminal type, `docker-compose stop cloud-code` to stop the current container
3. In your terminal type, `docker-compose build cloud-code` to build the new container with your current changes
4. In your terminal type, `docker-compose up -d cloud-code` to start the recently built container

If you change something in `docker-compose.yml`, you need to:
1. In your terminal type, `docker-compose restart parse cloud-code` to restart the containers

**Reply by cbaker6** [reply]
If you update to [0.8.4](https://github.com/netreconlab/parse-server-swift/releases/tag/0.8.4) it will automatically delete the old hooks when you restart ParseServerSwift (in the linux container or Xcode) which will fix part of the issue I mentioned in https://github.com/netreconlab/parse-server-swift/discussions/30#discussioncomment-6020232 as it seems like you are not deleting old hooks leftover from your XCTests. You should be sure to always verify the `Destination` column in the `Webhooks` section of your `Parse Dashboard` to ensure it's pointing to the correct ParseServerSwift endpoint you are testing/using

**Reply by jaysonng** [reply]
So I finally got it working when I finally pinpointed what that 127.0.0.1 error was.

Apparently my localhost isn't pointing to 127.0.0.1 ? Anyway, long story short, 

http://localhost:8081 != http://127.0.0.1/8081 which for some reason is what node.js is using instead of localhost. 

so in .env I changed PARSE_SERVER_SWIFT_HOST_NAME from localhost  to 127.0.0.1 and got it playgrounds to work finally.

Docker setup also worked before this revelation where I didn't realize that I had to delete all triggers pointing to localhost. I thought i only needed to delete or change the hello webhook function I was testing.  Your 0.8.4 update worked like a charm.

thanks @cbaker6, I can finally go about creating some Swift Cloud Code.

I'll just figure out the localhost not pointing to 127.0.0.1 problem or why Parse-Server node.js is looking for it next time.

**Reply by cbaker6** [reply]
> So I finally got it working when I finally pinpointed what that 127.0.0.1 error was.

I'm glad you got your setup working!

> I can finally go about creating some Swift Cloud Code.

You should be able to write Cloud Code fast since you are already familiar with ParseSwift. Definitely let me know if run into any bugs or features you feel are missing.

> http://localhost:8081/ != http://127.0.0.1/8081 which for some reason is what node.js is using instead of localhost.

I've never seen `localhost` and `127.0.0.1` not be localhost, maybe your computer is force using IPv6 and not allowing IPv4 at all? https://www.lifewire.com/network-computer-special-ip-address-818385#toc-how-127001-works

**Reply by jaysonng** [reply]
> I've never seen localhost and 127.0.0.1 not be localhost, maybe your computer is force using IPv6 and not allowing IPv4 at all? https://www.lifewire.com/network-computer-special-ip-address-818385#toc-how-127001-works

Right? I even checked /etc/hosts and it does say localhost is 127.0.0.1

Ill have to check on the link you sent in a couple of days. 

Thanks again!

**Comment by cbaker6** [comment]
Sidenote: you can create a simple CloudCode function (similar to the HelloWorld example) to return `"\(try await ParseServer.information())"` to solve your the problem you posted: https://community.parseplatform.org/t/parse-server-version-property/3441/7
